flocker/node/agents/blockdevice.py

```diff
@@ -1308,6 +1460,98 @@ def shared_state_changes(self):
         )
 
 
+# Mapping from desired and discovered dataset state to
+# IStateChange factory. (The factory is expected to take
+# ``desired_dataset`` and ``discovered_dataset``.
+Desired = Discovered = DatasetStates
+DO_NOTHING = lambda **kwargs: NoOp()
+DATASET_TRANSITIONS = {
+    Desired.MOUNTED: {
+        Discovered.NON_EXISTENT:
+            CreateBlockDeviceDataset.from_state_and_config,
+        # Other node will need to deatch first
+        Discovered.ATTACHED_ELSEWHERE: DO_NOTHING,
+        Discovered.ATTACHED_NO_FILESYSTEM:
+            CreateFilesystem.from_state_and_config,
+        Discovered.NON_MANIFEST: AttachVolume.from_state_and_config,
```

Comment:
I don't *think* that case should ever occur here. That case is we have stale non-manifest data. But we now never actually consult external state (except for applications).